DOC add page on curve sampling and stopping criterion (#529)
benchopt/benchopt: doc/convergence_curves.rst

.. _convergence_curves:

Sampling of convergence curves: ``stopping_strategy``
=====================================================

There are two ways to create a convergence curve for a given solver, controlled by the ``stopping_strategy`` attribute of the solver.

1. Using iterations or tolerance
--------------------------------

The first way, used for black box solvers, is to create curves by calling the ``Solver.run`` method several times for each solver with different values for the ``stop_val`` parameter.
The solver is restarted from scratch at each call to ``run``.

> - if the solver's ``stopping_strategy`` is ``"iteration"``, the number of iterations passed to ``run`` increases geometrically from 1 with a factor XXX, eventually switching to XXX if the curves becomes flat., and the

Reviewer: Thanks for the great PR ! Could it be explained how exactly is the curve deemed flat with pointers to the lines of code that take care of that ?
I am assuming it uses heuristics similar to this dynamic learning rate scheduler (http://blog.dlib.net/2018/02/automatic-learning-rate-scheduling-that.html) ?

Author: @jeandut I did another pass on the PR, could you point anything that remains unclear? Thanks for the great feedback :pray:

Reviewer: Done !

.. _convergence_curves:

Sampling of the convergence curves
==================================

For each solver, there are two ways to create a convergence curve.
They are chosen by the ``stopping_strategy`` attribute of the solver.

1. Using iterations or tolerance
--------------------------------

The first way is to use ``Solver.stopping_strategy = "iteration"`` or ``Solver.stopping_strategy = "tolerance"``.

> This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or numerical tolerance.

Reviewer 2: ```suggestion
This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or for a given numerical tolerance.
```

> This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or numerical tolerance.

Reviewer 2: ```suggestion
This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or for a given numerical tolerance.
```

This stopping strategy creates curves by calling ``Solver.run(stop_val)`` several times with different values for the ``stop_val`` parameter:

- if the solver's ``stopping_strategy`` is ``"iteration"``, ``stop_val`` is  the number of iterations passed to ``run``.

>   It increases geometrically by at least 1, starting from 1 with a factor 1.5.

Reviewer: Maybe a small markdown equation would help ? But this is already super helpful.

  Note that the first call uses a number of iterations of 0.

- if the solver's ``stopping_strategy`` is ``"tolerance"``, the ``stop_val`` parameter corresponds to the numerical tolerance.
  It decreases geometrically by a factor of 1.5 between each call to ``run``, starting from 1 at the second call.
  Note that the first call uses a tolerance of 1e38.

> In both cases, if the objective curve is flat (i.e., the variation of the objective between two points is numerically 0), the geometric rate is multiplied by 1.2.

Reviewer: Thank you ! Very helpful !

> Note that the solver is started from scratch at each call to ``solver.run``.

Reviewer: ```suggestion
Note that the solver is restarted from scratch at each call to ``solver.run``.
```

Reviewer 2: I would move this comment up, before describing the two strqtegy, to avoid people missing it, considering that it is part of the second strategy. Or it could be highlighted as a `note`.

For more advanced configurations, the evolution of ``stop_val`` can be controlled on a per solver basis, by implementing a static  ``Solver.get_next`` method, which receives the current value for tolerance/number of iterations, and returns the next one.

2. Using a callback
-------------------

> restarting the solver from scratch, though inevitable to handle black box solvers, can be costly.

Reviewer: ```suggestion
Restarting the solver from scratch, though inevitable to handle black box solvers, can be costly.
```

When a solver exposes the intermediate values of the iterates, it is possible to create the curve in a single solver run, by using ``stopping_strategy = "callback"``.
In that case, the argument passed to ``Solver.run`` will be a callable object, ``callback``.

> Every time it is called inside ``Solver.run`` with the current iterate ``x`` as argument, this object will store the objective value at the current iterate, and the current running time.

Reviewer 2: It is not this.
Actually, `stopping_strategy == "callback"` follows the same principle as `"iteration"`, except that `n_iter` is replaced by the number of call to the callback.

The loss is computed only when reaching this `stop_val` iterations and `shoud_stop` is also only called at this moment.

I think I agree with @jeandut that this is a bit a "melange des genres". I am all up to improve this, but the best would be to take time to discuss what we wanna do, to have a proper solution. Happy to chat IRL with you two.

Author: I don't get it, there is a single loss computed per run, which is the final one?

Reviewer 2: No, the loss is computed at the same point where it would have been with `stopping_strategy == "iteration"`, and the `should_stop` is also evaluated at the same points.
Basically, `callback` is a mirror for `iteration`, where the number of iterations is equal to the number of call to the callback.

Reviewer 2: A small pass on this text.
```suggestion
Like with ``stopping_strategy == "iteration"``, the loss is computed after a number of callback's calls that grows geometrically. If the loss was computed after :math:`n` calls, the loss and timing will be computed again when reaching :math:`\max(n+1, \rho * n)` calls to the callback.
The callback makes sure we do not account for loss computation time and also check for convergence every time the loss is computed (as described in the next section).
It returns ``False`` when the solver should be stopped. A classical usage pattern is:

.. code:: python

    def run(self, callback):
        x = ... # Initialize iterate

        while callback(x):
            x = ...  # Update iterate
        self.x = x

```

Each call to ``callback`` also calls  ``Solver.StoppingCriterion.should_stop`` and returns a boolean indicating whether or not the ``StoppingCriterion`` deems the solver should be stoppped.

When are the solvers stopped?
=============================

> For each of the sampling strategies above, the solvers continue running (i.e. the callback returns ``True``, the number of iterations passed to ``Solver.run`` increases or the tolerance passed to ``Solver.run`` decreases) until ``Solver.StoppingCriterion.should_stop()`` returns ``True``.
> 
> This value is determined by ``Solver.StoppingCriterion.check_convergence()``, based on the objective curve so far.
> For example, the ``check_convergence`` method of a ``SufficientDescentCriterion(eps, patience)`` returns ``True`` when the relative decrease of the objective was less than a tolerance ``eps`` for more than ``patience`` calls to ``check_convergence``.
> 

Reviewer 2: ```suggestion
For each of the sampling strategies above, the solvers continue running (i.e. the callback returns ``True``, the number of iterations passed to ``Solver.run`` increases or the tolerance passed to ``Solver.run`` decreases) until it the ``StoppingCriterion.should_stop()`` associated to the solver ``Solver.stopping_criterion`` returns ``True``.

This method takes into account the maximal number of runs given as ``--max-runs``, the timeout given by ``--timeout`` and also tries to stop the solver if it has converged.
The convergence of a solver is determined by  the ``StoppingCriterion.check_convergence()`` method, based on the objective curve so far.
There are three ``StoppingCriterion`` implemented in ``benchopt``:
- ``SufficientDescentCriterion(eps, patience)`` considers that the solver has converged when the relative decrease of the objective was less than a tolerance ``eps`` for more than ``patience`` calls to ``check_convergence``.
- ``SufficientProgressCriterion(eps, patience)`` considers that the solver has converged when the objective has not decreased by more than a tolerance ``eps`` for more than ``patience`` calls to ``check_convergence``.
- ``SingleRunCriterion(stop_val)`` only call the solver once with the given stop_val. This criterion designed for methods that converge to a given value, when one aim to benchmark final performance of multiple solvers.

```

.. _convergence_curves:

How are the convergence curves sampled?
=======================================

For each solver, there are two ways to create a convergence curve.
They are chosen by the ``stopping_strategy`` attribute of the solver.

1. Using iterations or tolerance
--------------------------------

The first way is to use ``Solver.stopping_strategy = "iteration"`` or ``Solver.stopping_strategy = "tolerance"``.
This is used for black box solvers, where one can only get the result of the solver for a given number of iterations or for a given numerical tolerance.
This stopping strategy creates curves by calling ``Solver.run(stop_val)`` several times with different values for the ``stop_val`` parameter:

- if the solver's ``stopping_strategy`` is ``"iteration"``, ``stop_val`` is the number of iterations passed to ``run``.

>   It increases geometrically by at least 1, starting from 1 with a factor 1.5.

Reviewer 2: ```suggestion
  It increases geometrically by at least 1, starting from 1 with a factor :math:`\rho=1.5`.
```

  Note that the first call uses a number of iterations of 0.
  The value from one call to the other follows:

  .. math::

    \text{stop_val} = \max(\text{stop_val} + 1, \text{int}(\rho * \text{stop_val}))

- if the solver's ``stopping_strategy`` is ``"tolerance"``, the ``stop_val`` parameter corresponds to the numerical tolerance.

>   It decreases geometrically by a factor of 1.5 between each call to ``run``, starting from 1 at the second call.

Reviewer 2: ```suggestion
  It decreases geometrically by a factor :math:`\rho=1.5` between each call to ``run``, starting from 1 at the second call.
```

2. Using a callback
-------------------

Restarting the solver from scratch, though inevitable to handle black box solvers, can be costly.

When a solver exposes the intermediate values of the iterates, it is possible to create the curve in a single solver run, by using ``stopping_strategy = "callback"``.
In that case, the argument passed to ``Solver.run`` will be a callable object, ``callback``.
Like with ``stopping_strategy == "iteration"``, the loss is computed after a number of callback's calls that grows geometrically. If the loss was computed after :math:`n` calls, the loss and timing will be computed again when reaching :math:`\max(n+1, \rho * n)` calls to the callback.
The callback makes sure we do not account for loss computation time and also check for convergence every time the loss is computed (as described in the next section).
It returns ``False`` when the solver should be stopped. A classical usage pattern is:

.. code:: python

    def run(self, callback):
        x = ... # Initialize iterate

        while callback(x):
            x = ...  # Update iterate
        self.x = x

> 
> Each call to ``callback`` also calls  ``Solver.StoppingCriterion.should_stop`` and returns a boolean indicating whether or not the ``StoppingCriterion`` deems the solver should be stoppped.
> 

Reviewer 2: ```suggestion
```
